chunky/src/java/se/llbit/chunky/world/Chunk.java

```diff
@@ -422,56 +438,54 @@ public synchronized void renderTopography() {
   /**
    * Load the block data for this chunk.
    *
-   * @param blocks block order: y, z, x.
+   * @param chunkData to be reused, if null one is created
+   * @param palette
    */
-  public synchronized void getBlockData(int[] blocks, byte[] biomes,
-      Collection<CompoundTag> tileEntities, Collection<CompoundTag> entities,
-      BlockPalette blockPalette) {
-
-    for (int i = 0; i < CHUNK_BYTES; ++i) {
-      blocks[i] = blockPalette.airId;
-    }
-
-    for (int i = 0; i < X_MAX * Z_MAX; ++i) {
-      biomes[i] = 0;
-    }
-
+  public synchronized ChunkData getChunkData(ChunkData chunkData, BlockPalette palette) {
     Set<String> request = new HashSet<>();
     request.add(DATAVERSION);
     request.add(LEVEL_SECTIONS);
     request.add(LEVEL_BIOMES);
     request.add(LEVEL_ENTITIES);
     request.add(LEVEL_TILEENTITIES);
     Map<String, Tag> data = getChunkData(request);
+    if(chunkData == null) {
+      chunkData = new GenericChunkData();
+    } else {
+      chunkData.clear();
+    }
     // TODO: improve error handling here.
     if (data == null) {
-      return;
+      return chunkData;
     }
     Tag sections = data.get(LEVEL_SECTIONS);
     Tag biomesTag = data.get(LEVEL_BIOMES);
     Tag entitiesTag = data.get(LEVEL_ENTITIES);
     Tag tileEntitiesTag = data.get(LEVEL_TILEENTITIES);
     if (biomesTag.isByteArray(X_MAX * Z_MAX) || biomesTag.isIntArray(X_MAX * Z_MAX)) {
-      extractBiomeData(biomesTag, biomes);
+      extractBiomeData(biomesTag, chunkData);
     }
 
     if (sections.isList()) {
-      loadBlockData(data, blocks, blockPalette);
+      loadBlockData(data, chunkData, palette);
 
       if (entitiesTag.isList()) {
+        Collection<CompoundTag> entities = chunkData.getEntities();
```

Comment:
This looks like it already returns the entity list even if it's just used to add entities to. Maybe change it to `chunkData.setEntities(((ListTag) entitiesTag).stream().filter(Tag::isCompountTag) …` to make clear that we are _setting_ the entities list here?

Reply:
Same for the tileEntities list. Not sure about using streams though.

Reply:
I was also debating having `addEntity` and `addTileEntity` methods, as well as the getter, just for clarification purposes